File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):    
    d = [[1,0], [-1,0], [0,1], [0,-1]]
    
    q = deque([[0, 0]])
    
    while q:
        x, y = q.popleft()
        move = maps[y][x]

        for dx, dy in d:
            nx = x + dx
            ny = y + dy

            if nx < 0 or nx >= len(maps[0])\
                or ny < 0 or ny >= len(maps): 
                continue
            
            if maps[ny][nx] == 1:
                maps[ny][nx] = move + 1
                q.append([nx, ny])

    return maps[-1][-1] if maps[-1][-1] != 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    private int n = 0;
    private int m = 0;
    private int[] dx = new int[]{1, 0, -1, 0};
    private int[] dy = new int[]{0, 1, 0, -1};
    
    public int solution(int[][] maps) {
        n = maps[0].length;
        m = maps.length;
                
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        
        while (!queue.isEmpty()) {
            int[] data = queue.poll();
            int x = data[0];
            int y = data[1];
            
            int move = maps[y][x];
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    continue;
                }
                
                if (maps[ny][nx] != 1) {
                    continue;
                }
                
                if (nx == n-1 && ny == m-1) {
                    return move + 1;
                }
                
                maps[ny][nx] = move + 1;
                queue.offer(new int[]{nx, ny});
            }
        }
        
        return -1;
    }
}